File: Algorithm/Kruskal's Algorithm/atcoder/A67 - MST (Minimum Spanning Tree)/A67.php
<?php

fscanf(STDIN, "%d %d", $N, $M);
$edges = [];

for ($i = 0; $i < $M; $i++) {
    fscanf(STDIN, "%d %d %d", $a, $b, $c);
    // 0-indexed に変換
    $edges[] = [$c, $a - 1, $b - 1];
}

// 重みで昇順にソート
usort($edges, fn($e1, $e2) => $e1[0] <=> $e2[0]);

// Union-Find 構造体
class UnionFind {
    private array $parent;
    private array $rank;

    public function __construct(int $n) {
        $this->parent = range(0, $n - 1);
        $this->rank = array_fill(0, $n, 0);
    }

    public function find(int $x): int {
        if ($this->parent[$x] !== $x) {
            $this->parent[$x] = $this->find($this->parent[$x]); // 経路圧縮
        }
        return $this->parent[$x];
    }

    public function union(int $x, int $y): bool {
        $rx = $this->find($x);
        $ry = $this->find($y);
        if ($rx === $ry) return false;

        if ($this->rank[$rx] < $this->rank[$ry]) {
            $this->parent[$rx] = $ry;
        } else {
            $this->parent[$ry] = $rx;
            if ($this->rank[$rx] === $this->rank[$ry]) {
                $this->rank[$rx]++;
            }
        }
        return true;
    }
}

$uf = new UnionFind($N);
$total = 0;
$count = 0;

foreach ($edges as [$cost, $u, $v]) {
    if ($uf->union($u, $v)) {
        $total += $cost;
        $count++;
        if ($count === $N - 1) break;
    }
}

echo $total . PHP_EOL;

File: Algorithm/Kruskal's Algorithm/atcoder/A67 - MST (Minimum Spanning Tree)/A67.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strconv"
)

type Edge struct {
	u, v, cost int
}

type UnionFind struct {
	parent []int
	rank   []int
}

func NewUnionFind(n int) *UnionFind {
	parent := make([]int, n)
	rank := make([]int, n)
	for i := 0; i < n; i++ {
		parent[i] = i
	}
	return &UnionFind{parent, rank}
}

func (uf *UnionFind) Find(x int) int {
	if uf.parent[x] != x {
		uf.parent[x] = uf.Find(uf.parent[x]) // 経路圧縮
	}
	return uf.parent[x]
}

func (uf *UnionFind) Union(x, y int) bool {
	rx := uf.Find(x)
	ry := uf.Find(y)
	if rx == ry {
		return false
	}
	if uf.rank[rx] < uf.rank[ry] {
		uf.parent[rx] = ry
	} else {
		uf.parent[ry] = rx
		if uf.rank[rx] == uf.rank[ry] {
			uf.rank[rx]++
		}
	}
	return true
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Split(bufio.ScanWords)

	nextInt := func() int {
		scanner.Scan()
		n, _ := strconv.Atoi(scanner.Text())
		return n
	}

	N := nextInt()
	M := nextInt()

	edges := make([]Edge, M)
	for i := 0; i < M; i++ {
		a := nextInt() - 1
		b := nextInt() - 1
		c := nextInt()
		edges[i] = Edge{a, b, c}
	}

	// コスト順にソート
	sort.Slice(edges, func(i, j int) bool {
		return edges[i].cost < edges[j].cost
	})

	uf := NewUnionFind(N)
	total := 0
	count := 0

	for _, e := range edges {
		if uf.Union(e.u, e.v) {
			total += e.cost
			count++
			if count == N-1 {
				break
			}
		}
	}

	fmt.Println(total)
}
